### ISSUE TITLE
Support thunks for argument default values

### ISSUE THREAD

**Issue by mnmtanish** [graphql/graphql-js#270]
It can be useful to support thunks to generate `defaultValue` for graphql arguments. Consider this example where the query time range is set dynamically. I'll send a pull-request if this is okay.

```
const BlogPosts = {
  type: new GraphQLList(BlogPost),
  args: {
    endTime: {
      type: GraphQLFloat,
      defaultValue: () => Date.now(),
    },
    startTime: {
      type: GraphQLFloat,
      defaultValue: () => Date.now() - 1000*60*60,
    },
  },
  async resolve(root, args) {
    // query blog posts ...
  },
};
```

I know I can do it inside the resolve function but for one of the graphql projects I'm working on (which is not a blog), we're re-using argument definitions. So the above example code would actually look something like this.

```
import {endTime, startTime} from './arguments';

const BlogPosts = {
  type: new GraphQLList(BlogPost),
  args: {endTIme, startTime},
  async resolve(root, args) {
    // query blog posts ...
  },
};
```

So, if it's *normal* to re-use argument definitions, supporting thunks for generating the `defaultValue` can be really useful to keep everything related to the argument in the same place.

**Possible Conflicts**

If ENUMs can have `function` values, it may not be possible to use functions for the `defaultValue`.

**Comment by leebyron** [comment]
I do not think we should support this because I think there are very few cases where this adds value, and there are considerable complications it provides. It also can present some fairly serious foot-guns. Actually the example you specify here I believe is a foot-gun.

This specific case came up before in #240 

There are two problems to consider here. 

1. defaultValue should be a constant value, so that it may be serialized and provided to the client so clients can determine if they do not need to provide a value or not. Dynamic default values like this would change every time you query introspection to retrieve the schema. 

2. Time is almost always the example brought up for this kind of thing, but there are two pretty critical issues with using a dynamically generated default value for Time. The first is that there is a time delta between determining the values for variables and resolving a field. This can potentially cause issues. `Date.now()` has a different value every time you call it. If your program is sensitive to time, you need to consider this carefully - perhaps you want to compute `Date.now()` one time before your GraphQL query starts to execute, as an example. The second more important concern is time shift between your client and your server. Any time a server is providing default time values on behalf of a client, you need to be cautious of shifting time. Clients often have wildly differing clock time from servers and this can result in all kinds of weird behavior.

TL;DR, I'm not sure if we want to open this can of worms as an explicit feature.

**Comment by mnmtanish** [comment]
Got it. Thanks for the comment
